### PR TITLE
Ft/zenko 144 https proxy

### PR DESCRIPTION
**Pull request by dora-korpar**
Adds https proxy and ssl certificates compatibility.
Proxy can be set either through the HTTP_PROXY or HTTPS_PROXY env variables
OR
In config.json with the following format:
"outboundProxy:" {
    "url": "http://myproxy.host:3129",    // required
    "key": "path/to/key",                         // optional
    "cert": "path/to/cert",                       // optional
    "caBundle": "path/to/ca"                 // optional
}

Here's the link to the squid Docker image on DockerHub: https://hub.docker.com/r/scality/ci-squid/

### CODE REVIEW

**Comment by dora-korpar** [comment]
@ironman-machine 
DEFAULT_BRANCH=rel/7.4-beta
try

File: lib/data/external/AzureClient.js [hunk 1]
```diff
@@ -22,12 +23,28 @@ class AzureClient {
             this._azureStorageEndpoint);
         this._dataStoreName = config.dataStoreName;
         this._bucketMatch = config.bucketMatch;
-        if (config.proxy) {
-            const parsedUrl = url.parse(config.proxy);
+        if (config.proxy && config.proxy.url) {
```
Comment (jonathan-gramain) [inline lib/data/external/AzureClient.js]: Ideally we could have a few helper methods or classes to do this prep job in an homogeneous manner for backends that support a proxy. I'm good with this though considering the time constraints.

File: lib/data/locationConstraintParser.js [hunk 2]
```diff
@@ -41,23 +41,35 @@ function parseLC() {
...
+                    options.key = keypath ?
+                        fs.readFileSync(keypath, 'ascii') : undefined;
+                    options.ca = capath ?
+                        fs.readFileSync(capath, 'ascii') : undefined;
+                    options.secureProxy = true;
+                }
+                options.keepAlive = true;
+                // whether options.secureProxy is set to true determines
+                // https or http proxy
+                connectionAgent = new HttpsProxyAgent(options);
```
Comment (jonathan-gramain) [inline lib/data/locationConstraintParser.js]: Does the construction and use of `HttpsProxyAgent` have a chance to work if `config.outboundProxy.certFilePaths` is not defined? In such case we could fall back to `https.Agent` or invalidate the config? Not sure just wondering.

File: lib/Config.js [hunk 3]
```diff
@@ -739,37 +744,69 @@ class Config extends EventEmitter {
...
+                ca: certPathsObj.capath,
+                cert: certPathsObj.certpath,
             };
-        } else if (key || cert) {
-            throw new Error('bad config: both certFilePaths.key and ' +
-                'certFilePaths.cert must be defined');
+        }
+
+        this.outboundProxy = {};
+        const envProxy = process.env.HTTP_PROXY || process.env.HTTPS_PROXY;
```
Comment (jeremyds) [inline lib/Config.js]: as seen with you, may be useful to handle their lowercase equivalents

File: squid/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,11 @@
+FROM alpine:3.7
```
Comment (jonathan-gramain) [inline squid/Dockerfile]: There may be a better place to put the `squid` directory than at the root, e.g. as a subdir of a `deps` or `external` folder.

File: lib/data/utils.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    // noProxy env var is a comma separated list of urls not to proxy
+    const noProxyList = noProxy.split(',');
+    if (noProxyList.includes(endpoint)) {
+        return true;
+    }
+    const epArr = endpoint.split('.');
+    // reverse array to make comparison easier
+    epArr.reverse();
+    let match = false;
+    for (let j = 0; j < noProxyList.length; j++) {
```
Comment (jonathan-gramain) [inline lib/data/utils.js]: You may want to look at this for later refinement: https://github.com/isaacs/minimatch

**Comment by dora-korpar** [comment]
@ironman-machine try

File: lib/data/locationConstraintParser.js [hunk 6]
```diff
@@ -41,26 +42,39 @@ function parseLC() {
...
             const sslEnabled = locationObj.details.https === true;
+            const signatureVersion = !sslEnabled ? 'v2' : 'v4';
             // TODO: HTTP requests to AWS are not supported with V4 auth for
             // non-file streams which are used by Backbeat. This option will be
             // removed once CA certs, proxy setup feature is implemented.
-            const signatureVersion = !sslEnabled ? 'v2' : 'v4';
-            const endpoint = locationObj.type === 'gcp' ?
-                locationObj.details.gcpEndpoint :
-                locationObj.details.awsEndpoint;
+            // const signatureVersion = !sslEnabled ? 'v2' : 'v4';
```
Comment (RachedBenMustapha) [inline lib/data/locationConstraintParser.js]: this looks like a leftover

**Comment by dora-korpar** [comment]
@ironman-machine try

File: lib/data/locationConstraintParser.js [hunk 7]
```diff
@@ -41,26 +42,39 @@ function parseLC() {
...
+                locationObj.details.gcpEndpoint :
+                locationObj.details.awsEndpoint;
+            const proxyMatch = proxyCompareURL(endpoint);
+            if (config.outboundProxy.url && !proxyMatch) {
+                const options = url.parse(config.outboundProxy.url);
+                if (config.outboundProxy.certFilePaths) {
+                    const { keypath, certpath, capath } =
+                        config.outboundProxy.certFilePaths;
+                    options.cert = certpath ?
+                        fs.readFileSync(certpath, 'ascii') : undefined;
```
Comment (rahulreddy) [inline lib/data/locationConstraintParser.js]: Same thing, if you read the certs into an object in Config, you don't have to redo this in multiple places.

**Comment by dora-korpar** [comment]
@ironman-machine try

**Pull request merged**